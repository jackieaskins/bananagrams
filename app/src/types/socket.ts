import { BoardLocation } from "./board";
import { Game } from "./game";
import { PlayerStatus } from "./player";

export enum ClientToServerEventName {
  CreateGame = "createGame",
  JoinGame = "joinGame",
  KickPlayer = "kickPlayer",
  LeaveGame = "leaveGame",
  SetStatus = "setStatus",
  Split = "split",
  Peel = "peel",
  Dump = "dump",
  MoveTileFromHandToBoard = "moveTileFromHandToBoard",
  MoveTileFromBoardToHand = "moveTileFromBoardToHand",
  MoveTileOnBoard = "moveTileOnBoard",
  ShuffleHand = "shuffleHand",
}

export enum ServerToClientEventName {
  Notification = "notification",
  GameInfo = "gameInfo",
}

export type SocketCallback<T> = (
  error: { message: string } | null,
  args: T | null,
) => void;

export type SocketEventHandler<Event = null, CallbackArg = void> = (
  event: Event,
  callback?: SocketCallback<CallbackArg>,
) => void;

export type ClientToServerEvents = {
  [ClientToServerEventName.CreateGame]: SocketEventHandler<
    { gameName: string; username: string; isShortenedGame: boolean },
    Game
  >;
  [ClientToServerEventName.JoinGame]: SocketEventHandler<
    { gameId: string; username: string; isSpectator: boolean },
    Game
  >;
  [ClientToServerEventName.LeaveGame]: SocketEventHandler;

  [ClientToServerEventName.KickPlayer]: SocketEventHandler<{ userId: string }>;
  [ClientToServerEventName.SetStatus]: SocketEventHandler<{
    status: PlayerStatus;
  }>;

  [ClientToServerEventName.Split]: SocketEventHandler;
  [ClientToServerEventName.Peel]: SocketEventHandler;
  [ClientToServerEventName.Dump]: SocketEventHandler<{
    tileId: string;
    boardLocation: BoardLocation | null;
  }>;

  [ClientToServerEventName.ShuffleHand]: SocketEventHandler;

  [ClientToServerEventName.MoveTileFromHandToBoard]: SocketEventHandler<{
    tileId: string;
    boardLocation: BoardLocation;
  }>;
  [ClientToServerEventName.MoveTileFromBoardToHand]: SocketEventHandler<{
    boardLocation: BoardLocation;
  }>;
  [ClientToServerEventName.MoveTileOnBoard]: SocketEventHandler<{
    fromLocation: BoardLocation;
    toLocation: BoardLocation;
  }>;
};

export type ServerToClientEvents = {
  [ServerToClientEventName.Notification]: SocketEventHandler<{
    message: string;
  }>;
  [ServerToClientEventName.GameInfo]: SocketEventHandler<Game>;
};
