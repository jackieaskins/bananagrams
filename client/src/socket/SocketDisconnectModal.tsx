import { Button, Modal } from 'antd';
import { useCallback, useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';

import { addDisconnectListener } from '.';

const SocketDisconnectModal = (): JSX.Element => {
  const [shouldShowModal, setShouldShowModal] = useState(false);
  const { push } = useHistory();

  useEffect(() => {
    addDisconnectListener(() => {
      setShouldShowModal(true);
    });
  }, []);

  const hideModal = useCallback(() => {
    push('/');
    setShouldShowModal(false);
  }, [push]);

  return (
    <Modal
      closable={false}
      footer={<Button onClick={hideModal}>Return home</Button>}
      title="Disconnected"
      visible={shouldShowModal}
    >
      You have been disconnected from the server
    </Modal>
  );
};

export default SocketDisconnectModal;
