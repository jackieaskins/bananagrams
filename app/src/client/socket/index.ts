import { Socket, io } from "socket.io-client";
import { ClientToServerEvents, ServerToClientEvents } from "@/types/socket";

export const socket: Socket<ServerToClientEvents, ClientToServerEvents> = io({
  autoConnect: false,
});
