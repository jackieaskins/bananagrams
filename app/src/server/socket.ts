import { Server } from "socket.io";
import GameController from "./controllers/GameController";
import {
  ClientToServerEventName,
  ClientToServerEvents,
  ServerToClientEvents,
  SocketCallback,
} from "@/types/socket";

export function handler<E, T>(fn: (event: E) => T) {
  return function (event: E, callback?: SocketCallback<T>): void {
    try {
      const returnValue = fn(event) || null;
      callback?.(null, returnValue);
    } catch (e) {
      const { message } = e as Error;
      callback?.({ message }, null);
    }
  };
}

export function configureSocket(
  io: Server<ClientToServerEvents, ServerToClientEvents>,
): void {
  io.on("connection", (socket) => {
    const { id: userId } = socket;

    console.log(`New connection: ${userId}`);

    let gameController: GameController | undefined;

    function withController<E, T>(
      fn: (event: E, controller: GameController) => T,
    ) {
      return handler((event: E) => {
        if (!gameController) {
          throw new Error("Not in a game");
        }

        return fn(event, gameController);
      });
    }

    socket.on(
      ClientToServerEventName.CreateGame,
      handler(({ gameName, username, isShortenedGame }) => {
        gameController = GameController.createGame(
          gameName,
          username,
          isShortenedGame,
          io,
          socket,
        );
        return gameController.getCurrentGame().toJSON();
      }),
    );

    socket.on(
      ClientToServerEventName.JoinGame,
      handler(({ gameId, username, isSpectator }) => {
        gameController = GameController.joinGame(
          gameId,
          username,
          isSpectator,
          io,
          socket,
        );
        return gameController.getCurrentGame().toJSON();
      }),
    );

    socket.on(
      ClientToServerEventName.KickPlayer,
      withController(({ userId }, controller) => controller.kickPlayer(userId)),
    );

    socket.on(
      ClientToServerEventName.LeaveGame,
      withController((_event, controller) => {
        controller.leaveGame();
        gameController = undefined;
      }),
    );

    socket.on(
      ClientToServerEventName.SetStatus,
      withController(({ status }, controller) => controller.setStatus(status)),
    );

    socket.on(
      ClientToServerEventName.Split,
      withController((_event, controller) => controller.split()),
    );

    socket.on(
      ClientToServerEventName.Peel,
      withController((_event, controller) => controller.peel()),
    );

    socket.on(
      ClientToServerEventName.Dump,
      withController(({ tileId, boardLocation }, controller) =>
        controller.dump(tileId, boardLocation),
      ),
    );

    socket.on(
      ClientToServerEventName.MoveTilesFromHandToBoard,
      withController(({ tiles }, controller) =>
        controller.moveTilesFromHandToBoard(tiles),
      ),
    );

    socket.on(
      ClientToServerEventName.MoveTilesFromBoardToHand,
      withController(({ boardLocations }, controller) =>
        controller.moveTilesFromBoardToHand(boardLocations),
      ),
    );

    socket.on(
      ClientToServerEventName.MoveTilesOnBoard,
      withController(({ locations }, controller) =>
        controller.moveTilesOnBoard(locations),
      ),
    );

    socket.on(
      ClientToServerEventName.ShuffleHand,
      withController((_event, controller) => controller.shuffleHand()),
    );

    socket.on("disconnect", () => {
      gameController?.leaveGame();
      gameController = undefined;
      console.log(`${userId} disconnected`);
    });
  });
}
