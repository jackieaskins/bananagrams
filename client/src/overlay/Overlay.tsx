import { Suspense } from 'react';
import Draggable from 'react-draggable';
import { useRecoilCallback } from 'recoil';

import LoadingContainer from '../loading/LoadingContainer';
import { moveTile } from '../socket';
import { selectedTileState } from '../tile/selectedTileState';
import Controls from './Controls';
import Hand from './LazyHand';

const Overlay = (): JSX.Element => {
  const handleClick = useRecoilCallback(
    ({ set, snapshot }) => async ({ target }) => {
      if (target.className.split(' ').includes('tile')) {
        return;
      }

      const selectedTile = await snapshot.getPromise(selectedTileState);

      if (selectedTile) {
        const {
          tile: { id: tileId },
          boardPosition,
        } = selectedTile;
        moveTile({
          tileId,
          fromPosition: boardPosition,
          toPosition: null,
        });
        set(selectedTileState, null);
      }
    },
    []
  );

  return (
    <Draggable cancel=".no-drag">
      <div
        className="overlay"
        css={{
          position: 'absolute',
          backgroundColor: 'rgba(173, 216, 230, 0.5)',
          border: '1px solid rgb(173, 216, 230)',
          width: '70vw',
          bottom: 0,
          left: 0,
          right: 0,
          margin: '15px auto',
          borderRadius: '10px',
          display: 'flex',
          justifyContent: 'space-between',
          alignItems: 'space-around',
          padding: '15px',
          transition: 'width 0.5s, height 0.5s',
        }}
        onClick={handleClick}
      >
        <Suspense fallback={<LoadingContainer />}>
          <Hand />
          <Controls />
        </Suspense>
      </div>
    </Draggable>
  );
};

export default Overlay;
