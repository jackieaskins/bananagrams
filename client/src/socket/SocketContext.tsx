import { useSnackbar } from 'notistack';
import React, { createContext, useContext, useEffect } from 'react';
import { Socket } from 'socket.io-client';

import socket from './index';

export type Callback<T> = (error: Error | null, data: T) => void;

type SocketState = {
  socket: Socket;
};

export const SocketContext = createContext<SocketState>({
  socket,
});

export const SocketProvider: React.FC = ({ children }) => {
  const { enqueueSnackbar } = useSnackbar();

  useEffect(() => {
    socket.on('notification', ({ message }: { message: string }) => {
      enqueueSnackbar(message);
    });

    return (): void => {
      socket.emit('disconnect');
    };
  }, []);

  return (
    <SocketContext.Provider value={{ socket }}>
      {children}
    </SocketContext.Provider>
  );
};
SocketProvider.displayName = 'SocketProvider';

export const useSocket = (): SocketState => useContext(SocketContext);
