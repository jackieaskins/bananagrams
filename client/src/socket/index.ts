import { io } from 'socket.io-client';

import { BoardPosition, BoardSquare } from '../boards/types';
import { GameInfo } from '../games/types';
import { Hand } from '../hands/types';

export type Callback<T> = (error: { message: string } | null, data: T) => void;
export type Listener<T = void> = (data: T) => void;

export const socket = io();

export const getUserId = (): string => socket.id;

export const createGame = (
  props: {
    gameName: string;
    username: string;
    isShortenedGame: boolean;
  },
  callback: Callback<GameInfo>
): void => {
  socket.emit('createGame', props, callback);
};
export const joinGame = (
  props: { gameId: string; username: string },
  callback: Callback<GameInfo>
): void => {
  socket.emit('joinGame', props, callback);
};
export const setReady = (props: { isReady: boolean }): void => {
  socket.emit('ready', props);
};
export const moveTile = (props: {
  tileId: string;
  fromPosition: BoardPosition | null;
  toPosition: BoardPosition | null;
}): void => {
  socket.emit('moveTile', props);
};
export const kickPlayer = (props: { userId: string }): void => {
  socket.emit('kickPlayer', props);
};
export const leaveGame = (props: { gameId: string }): void => {
  socket.emit('leaveGame', props);
};

export const addDisconnectListener = (listener: Listener): void => {
  socket.on('disconnect', listener);
};

export const addListeners = (
  gameInfoListener: Listener<GameInfo>,
  boardSquareUpdateListener: Listener<{
    id: string;
    square: BoardSquare | null;
  }>,
  handUpdateListener: Listener<Hand>
): void => {
  socket.on('gameInfo', gameInfoListener);
  socket.on('boardSquareUpdate', boardSquareUpdateListener);
  socket.on('handUpdate', handUpdateListener);
};
export const removeListeners = (): void => {
  socket.off('gameInfo');
  socket.off('boardSquareUpdate');
  socket.off('handUpdate');
};

export const disconnect = (): void => {
  socket.disconnect();
};
