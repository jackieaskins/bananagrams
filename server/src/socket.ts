import { Server } from 'socket.io';
import GameController from './controllers/GameController';

export const handler = (
  fn: () => Record<string, any> | void,
  callback?: (
    error: { message: string } | null,
    data: Record<string, any> | null
  ) => void
): void => {
  try {
    const returnValue = fn() || null;
    callback?.(null, returnValue);
  } catch ({ message }) {
    callback?.({ message }, null);
  }
};

export const configureSocket = (io: Server): void => {
  io.on('connection', (socket) => {
    const { id: userId } = socket;

    console.log(`New connection: ${userId}`);

    let gameController: GameController | undefined;

    const validateGameControllerExists = (): void => {
      if (!gameController) {
        throw new Error('Not in a game');
      }
    };

    socket.on(
      'createGame',
      ({ gameName, username, isShortenedGame }, callback) => {
        handler(() => {
          gameController = GameController.createGame(
            gameName,
            username,
            isShortenedGame,
            io,
            socket
          );
          return gameController.getCurrentGame().toJSON();
        }, callback);
      }
    );

    socket.on('joinGame', ({ gameId, username }, callback) => {
      handler(() => {
        gameController = GameController.joinGame(gameId, username, io, socket);
        return gameController.getCurrentGame().toJSON();
      }, callback);
    });

    socket.on('kickPlayer', ({ userId }, callback) => {
      handler(() => {
        validateGameControllerExists();
        (gameController as GameController).kickPlayer(userId);
      }, callback);
    });

    socket.on('leaveGame', ({}, callback) => {
      handler(() => {
        validateGameControllerExists();
        (gameController as GameController).leaveGame();
        gameController = undefined;
      }, callback);
    });

    socket.on('ready', ({ isReady }, callback) => {
      handler(() => {
        validateGameControllerExists();
        (gameController as GameController).setReady(isReady);
      }, callback);
    });

    socket.on('peel', ({}, callback) => {
      handler(() => {
        validateGameControllerExists();
        (gameController as GameController).peel();
      }, callback);
    });

    socket.on('dump', ({ tileId, boardLocation }, callback) => {
      handler(() => {
        validateGameControllerExists();
        (gameController as GameController).dump(tileId, boardLocation);
      }, callback);
    });

    socket.on(
      'moveTileFromHandToBoard',
      ({ tileId, boardLocation }, callback) => {
        handler(() => {
          validateGameControllerExists();
          (gameController as GameController).moveTileFromHandToBoard(
            tileId,
            boardLocation
          );
        }, callback);
      }
    );

    socket.on('moveTileFromBoardToHand', ({ boardLocation }, callback) => {
      handler(() => {
        validateGameControllerExists();
        (gameController as GameController).moveTileFromBoardToHand(
          boardLocation
        );
      }, callback);
    });

    socket.on('moveAllTilesFromBoardToHand', ({}, callback) => {
      handler(() => {
        validateGameControllerExists();
        (gameController as GameController).moveAllTilesFromBoardToHand();
      }, callback);
    });

    socket.on('moveTileOnBoard', ({ fromLocation, toLocation }, callback) => {
      handler(() => {
        validateGameControllerExists();
        (gameController as GameController).moveTileOnBoard(
          fromLocation,
          toLocation
        );
      }, callback);
    });

    socket.on('shuffleHand', ({}, callback) => {
      handler(() => {
        validateGameControllerExists();
        (gameController as GameController).shuffleHand();
      }, callback);
    });

    socket.on('disconnect', () => {
      gameController?.leaveGame();
      gameController = undefined;
      console.log(`${userId} disconnected`);
    });
  });
};
