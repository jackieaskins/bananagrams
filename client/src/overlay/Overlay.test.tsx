import { shallow } from 'enzyme';
import { useRecoilCallback } from 'recoil';

import { moveTile } from '../socket';
import Overlay from './Overlay';

const mockUseRecoilCallback = useRecoilCallback as jest.Mock;
const mockSet = jest.fn();
const mockSnapshot = {
  getPromise: jest.fn(),
};
jest.mock('recoil', () => ({
  useRecoilCallback: jest.fn((fn) =>
    fn({ set: mockSet, snapshot: mockSnapshot })
  ),
}));

const mockMoveTile = moveTile as jest.Mock;
jest.mock('../socket', () => ({
  moveTile: jest.fn(),
}));

const mockSelectedTileState = 'mockSelectedTileState';
jest.mock('../tile/selectedTileState', () => ({
  selectedTileState: 'mockSelectedTileState',
}));

describe('<Overlay />', () => {
  const renderComponent = (propOverrides = {}) =>
    shallow(<Overlay {...propOverrides} />);

  it('renders hand and controls', () => {
    expect(renderComponent()).toMatchSnapshot();
  });

  describe('handleClick', () => {
    const handleClick = async (event = { target: { className: '' } }) => {
      renderComponent();
      await mockUseRecoilCallback.mock.results[0].value(event);
    };

    it('returns early if tile was clicked', async () => {
      await handleClick({ target: { className: 'class tile' } });

      expect(mockMoveTile).not.toHaveBeenCalled();
    });

    it('gets selected tile', async () => {
      await handleClick();

      expect(mockSnapshot.getPromise).toHaveBeenCalledWith(
        mockSelectedTileState
      );
    });

    it('does not move tile if no selected tile', async () => {
      await handleClick();

      expect(moveTile).not.toHaveBeenCalled();
    });

    describe('when tile is selected', () => {
      beforeEach(async () => {
        mockSnapshot.getPromise.mockReturnValue({
          tile: { id: 'A1', letter: 'A' },
          boardPosition: null,
        });
        await handleClick();
      });

      it('moves tile', () => {
        expect(mockMoveTile).toHaveBeenCalledWith({
          tileId: 'A1',
          fromPosition: null,
          toPosition: null,
        });
      });

      it('clears selected tileId', () => {
        expect(mockSet).toHaveBeenCalledWith(mockSelectedTileState, null);
      });
    });
  });
});
