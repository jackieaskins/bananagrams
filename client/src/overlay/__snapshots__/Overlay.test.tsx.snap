// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Overlay /> renders hand and controls 1`] = `
.emotion-0jv3q9u-Overlay {
  position: absolute;
  background-color: rgba(173, 216, 230, 0.5);
  border: 1px solid rgb(173, 216, 230);
  width: 70vw;
  bottom: 0;
  left: 0;
  right: 0;
  margin: 15px auto;
  border-radius: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-items: space-around;
  -webkit-box-align: space-around;
  -ms-flex-align: space-around;
  align-items: space-around;
  padding: 15px;
  -webkit-transition: width 0.5s,height 0.5s;
  transition: width 0.5s,height 0.5s;
}

<Draggable
  allowAnyClick={false}
  axis="both"
  bounds={false}
  cancel=".no-drag"
  defaultClassName="react-draggable"
  defaultClassNameDragged="react-draggable-dragged"
  defaultClassNameDragging="react-draggable-dragging"
  defaultPosition={
    Object {
      "x": 0,
      "y": 0,
    }
  }
  disabled={false}
  enableUserSelectHack={true}
  grid={null}
  handle={null}
  offsetParent={null}
  onDrag={[Function]}
  onMouseDown={[Function]}
  onStart={[Function]}
  onStop={[Function]}
  position={null}
  scale={1}
  transform={null}
>
  <div
    className="overlay emotion-0"
    onClick={[Function]}
  >
    <Suspense
      fallback={<LoadingContainer />}
    >
      <lazy />
      <Controls />
    </Suspense>
  </div>
</Draggable>
`;
